crates/autopilot/src/limit_order_quoter.rs

```diff
@@ -0,0 +1,94 @@
+use crate::database::Postgres;
+use anyhow::Result;
+use chrono::Duration;
+use model::{
+    order::OrderKind,
+    quote::{default_verification_gas_limit, OrderQuoteSide, QuoteSigningScheme, SellAmount},
+    signature::Signature,
+};
+use shared::order_quoting::{OrderQuoting, QuoteParameters};
+use std::sync::Arc;
+
+/// Background task which quotes all limit orders and sets the surplus_fee for each one
+/// to the fee returned by the quoting process. If quoting fails, the corresponding
+/// order is skipped.
+pub struct LimitOrderQuoter {
+    pub limit_order_age: Duration,
+    pub loop_delay: std::time::Duration,
+    pub quoter: Arc<dyn OrderQuoting>,
+    pub database: Postgres,
+}
+
+impl LimitOrderQuoter {
+    pub fn spawn(self) {
+        tokio::spawn(async move {
+            loop {
+                if let Err(err) = self.update().await {
+                    tracing::error!(?err, "failed to update limit order surplus");
+                }
+                tokio::time::sleep(self.loop_delay).await;
+            }
+        });
+    }
+
+    async fn update(&self) -> Result<()> {
+        let orders = self
+            .database
+            .limit_orders_with_outdated_fees(self.limit_order_age)
+            .await?;
+        for order in orders {
+            match self
+                .quoter
+                .calculate_quote(QuoteParameters {
+                    sell_token: order.data.sell_token,
+                    buy_token: order.data.buy_token,
+                    side: match order.data.kind {
+                        OrderKind::Buy => OrderQuoteSide::Buy {
+                            buy_amount_after_fee: order.data.buy_amount,
+                        },
+                        OrderKind::Sell => OrderQuoteSide::Sell {
+                            sell_amount: SellAmount::AfterFee {
+                                value: order.data.sell_amount,
```

Comment:
I think this should be:
```suggestion
                            sell_amount: SellAmount::BeforeFee {
                                value: order.data.sell_amount + order.data.fee_amount,
```

Reply:
The reasoning here is that we want to simulate what a trade looks like with the full `order.data.sell_amount + order.data.fee_amount`.

What is confusing for limit orders is that the "effective sell amount" for a limit order (i.e. what gets actually traded against an AMM) is not `order.sell_amount`, like it is for market orders, but instead `effective_sell_amount = order.sell_amount + order.fee_amount - surplus_fee`.

Reply:
https://github.com/cowprotocol/services/pull/714/commits/0d585da21a9e0dbd225438d700b448d8b4e7e95b